Add hidden `zenml clean` command for use in development / debugging

strickvl opened the pull request:
## Describe changes
I added the functionality for a `zenml clean` command which can be used for the purposes of resetting your profiles / environment.

If you confirm via the CLI dialogue, or pass in the `-y` flag, it will delete your local `.zen` folder (if you have one) as well as the global config folder.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/features/integrations) table.
- [ ] I have added tests to cover my changes.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

schustmi commented on src/zenml/cli/base.py:
```diff
@@ -64,18 +67,31 @@ def init(path: Optional[Path]) -> None:
+    if yes or confirm:
+        local_zen_repo_config = Path.cwd() / REPOSITORY_DIRECTORY_NAME
+        global_zen_config = get_global_config_directory()
+        if exists(str(local_zen_repo_config)):
```
It feels a little strange to me to have `exists`/`rmtree` without the `fileio.` prefix similar to (most of) the remaining code, but feel free to ignore

schustmi commented on src/zenml/cli/base.py:
```diff
@@ -64,18 +67,31 @@ def init(path: Optional[Path]) -> None:
+        if exists(str(local_zen_repo_config)):
+            rmtree(str(local_zen_repo_config))
+            declare(f"Deleted local ZenML config from {local_zen_repo_config}.")
+        if exists(global_zen_config):
```
I'm wondering if we should try to keep the user id as well as the analytics opt-in status when deleting the global config directory. Personally whenever I'm looking to clean out my local setup, what I'm actually trying to do is delete local stores (=artifact/metadata stores) as well as some of the profiles. But that would require far more flexibility from this function, such as specifying profiles that are ignored from deletion.
As long as @htahir1 is fine with deleting the entire config I'm also okay with merging this as-is to speed up your workflow

htahir1 commented on src/zenml/cli/base.py:
```diff
@@ -64,18 +68,37 @@ def init(path: Optional[Path]) -> None:
+            declare(f"Deleted global ZenML config from {global_zen_config}.")
+
+        fileio.makedirs(str(global_zen_config))
+        yaml_utils.write_yaml(str(config_yaml_path), config_yaml_data)
```
The way this works it wont actually delete the profiles, but still leave a reference to them in the `config.yaml`. Here is an example config file

```
activated_profile: zenbytes
analytics_opt_in: false
profiles:
  default:
    active_stack: default
    name: default
    store_type: local
    store_url: file:///home/hamza/.config/zenml/profiles/default
  legacy-repository-2dc8050f:
    active_stack: aws_kubeflow_stack
    name: legacy-repository-2dc8050f
    store_type: local
    store_url: file:///home/hamza/workspace/zenml_io/zenml/.zen
  mlflow-as-stack:
    active_stack: default
    name: mlflow-as-stack
    store_type: local
    store_url: file:///home/hamza/.config/zenml/profiles/mlflow-as-stack
  wandb-as-stack:
    active_stack: default
    name: wandb-as-stack
    store_type: local
    store_url: file:///home/hamza/.config/zenml/profiles/wandb-as-stack
  zenbytes:
    active_stack: aws_kubeflow_stack
    name: zenbytes
    store_type: local
    store_url: file:///home/hamza/.config/zenml/profiles/zenbytes
user_id: 1f51b5b2-bac9-4a5b-9433-4b04e8eff5b6
version: 0.7.2
```

If you restore the `profiles` key they will now point to folders that dont exist :-) 

The cleanest way to do this is to use the logic of what is happening in `delete_profile` here (https://github.com/zenml-io/zenml/blob/aac824bd2b1733dec20ef7c70af1013bdc549cc9/src/zenml/cli/config.py#L209) . You can simply loop over all the profiles and do `cfg.delete_profile(name)` to make sure the deletion is done through the same canonical function

strickvl replied:
The only thing we need to preserve is the `user_id`, if I'm not mistaken. What is the point in using the `cfg` to delete things when I can just reinstantiate and switch out the `user_id`. Is this going to cause telemetry issues? I'm pushing a solution that'd do it the way I thought made sense.

htahir1 commented on src/zenml/cli/base.py:
```diff
@@ -64,18 +68,37 @@ def init(path: Optional[Path]) -> None:
+        Repository.initialize(root=Path.cwd())
+
+    else:
+        declare("Aborting clean.")
```
I would generally also log.info exactly what is deleted (i.e. each folder). To be more verbose about the deletion.

strickvl replied:
Done.

htahir1 reviewed (changes requested):
I like this feature and had it as a ticket too! Nice to see the initiative here.  Left a few suggestions

strickvl commented on src/zenml/cli/base.py:
```diff
@@ -64,18 +68,37 @@ def init(path: Optional[Path]) -> None:
     if not yes:
-        _ = confirmation(
-            "This will completely delete all pipelines, their associated "
+        confirm = confirmation(
```
Right. Done.

strickvl commented on src/zenml/cli/base.py:
```diff
@@ -64,18 +68,37 @@ def init(path: Optional[Path]) -> None:
+            fileio.rmtree(str(local_zen_repo_config))
+            declare(f"Deleted local ZenML config from {local_zen_repo_config}.")
+        if fileio.exists(str(global_zen_config)):
+            config_yaml_path = global_zen_config / "config.yaml"
```
Done.

schustmi reviewed (changes requested):
I think this would work as is, but I can see a few issues with it:
- We're deleting the analytics opt-in status of the user as well, which might be unexpected
- Even if we "migrate" the analytics status similar to the user-id, we will have to make sure that we don't forget to include exceptions here for any additional attributes we add in the future (same for any directories that should not be deleted btw)

I would suggest a different approach instead, that errs on the side of caution and deletes only explicitly stuff we want to delete. In future releases if we forget to update this function, we might be left with some uncleaned files but at least in a state that didn't delete any important information.

How would my solution look like:
I would simply loop over all profiles in the global config and delete them. The deletion of a profile already deletes the directory associated to it.
```python
for profile_name in GlobalConfig().profiles:
     # this already works with the existing implementation unless the profile 
     # is the active profile, we would need to handle that somehow
     GlobalConfig().delete_profile(profile)
```

The clean method could then delete additional directories that we want cleaned, e.g. the `zenml_examples` if they were pulled?

schustmi commented on src/zenml/cli/base.py:
```diff
@@ -64,18 +68,48 @@ def init(path: Optional[Path]) -> None:
+            fileio.rmtree(str(global_zen_config))
+            declare(f"Deleted global ZenML config from {global_zen_config}.")
+
+        Repository.initialize(root=Path.cwd())
```
Forgot one small thing: I'm not sure it's necessary to reinitialize a repo within the `clean` command, I think we should leave that up to the user. (I could certainly see myself calling this from a directory which I don't intend to use for future development)

AlexejPenner replied:
+1 on this one

stefannica commented on src/zenml/cli/base.py:
```diff
@@ -64,18 +68,48 @@ def init(path: Optional[Path]) -> None:
+            declare(f"Deleted local ZenML config from {local_zen_repo_config}.")
+        if fileio.exists(str(global_zen_config)):
+            config_yaml_path = global_zen_config / CONFIG_FILE_NAME
+            config_yaml_data = yaml_utils.read_yaml(str(config_yaml_path))
```
you don't have to load the yaml file, you already have all the information in the GlobalConfiguration() singleton object, so you can just use something like `old_user_id = GlobalConfiguration().user_id` instead.

You should also preserve the `analytics_opt_in` value.

AlexejPenner reviewed (commented):
Just very tiny things - I do agree with the @schustmi that It might make sense to persist information like opt-in status

stefannica commented on src/zenml/cli/base.py:
```diff
@@ -64,18 +68,48 @@ def init(path: Optional[Path]) -> None:
+            config_yaml_path = global_zen_config / CONFIG_FILE_NAME
+            config_yaml_data = yaml_utils.read_yaml(str(config_yaml_path))
+            old_user_id = config_yaml_data["user_id"]
+            for dir_name in fileio.listdir(str(global_zen_config)):
```
I think you could `rmtree` the entire global configuration folder rather than going through subfolders one by one.

stefannica commented on src/zenml/cli/base.py:
```diff
@@ -64,18 +68,48 @@ def init(path: Optional[Path]) -> None:
+
+        Repository.initialize(root=Path.cwd())
+        if old_user_id:
+            new_config_yaml_data = yaml_utils.read_yaml(str(config_yaml_path))
+            new_config_yaml_data["user_id"] = old_user_id
+            yaml_utils.write_yaml(str(config_yaml_path), new_config_yaml_data)
```
I don't think I understand what this does: you're basically reading and re-writing the same file contents you had before the CLI command was run.

I think you have to remove the `profiles` field, otherwise you end up with the same profiles as before, but it would be even better to do this via the GlobalConfiguration class itself instead of manipulating raw yaml files here. Calling the GlobalConfiguration constructor with custom values will create a new object that, in the absence of a configuration yaml file, will act as if you're running ZenML for the first time (which is what you want). The only issue is that you have to preserve the `user_id` and `analytics_opt_in` values somehow, which might require some small alterations to the GlobalConfiguration.__init__ method to allow you to supply them as constructor args.

strickvl commented:
@stefannica @AlexejPenner @schustmi I've addressed all the above comments in my latest commits pushed to this branch.

I've also made the command hidden as I agree that probably we don't want users to be doing this unnecessarily. (LMK if it's too paternalistic to hide the command...)

@stefannica you'll want to check / make sure that the special option/flag I added in for your local deletion idea is what you wanted.

htahir1 reviewed (changes requested):
I like this logic now a lot but its critical to test whether the user_id and config stuff is the same when the config gets regenerated. We should add a test for this

schustmi commented on src/zenml/cli/base.py:
```diff
@@ -63,19 +66,116 @@ def init(path: Optional[Path]) -> None:
+      local (flag; default value = False): If you want to delete local metadata and artifact stores from the active stack.
     """
+    if local:
+        if yes:
+            _delete_local_artifact_metadata(True)
+        else:
+            _delete_local_artifact_metadata()
```
```suggestion
        _delete_local_artifact_metadata(force_delete=yes)
```

schustmi commented on src/zenml/cli/base.py:
```diff
@@ -63,19 +66,116 @@ def init(path: Optional[Path]) -> None:
+            and not is_remote(str(artifact_store_path))
+        ):
+            # delete all files inside those directories
+            for path in metadata_store_path.iterdir():
```
Can't we just delete the entire metadata dir here instead of looping over it?

strickvl replied:
@stefannica said that he just wanted to delete the files inside the directories but not the directories themselves, if I didn't mishear him.

schustmi replied:
Can you explain why @stefannica ?

schustmi commented on src/zenml/cli/base.py:
```diff
@@ -63,19 +66,116 @@ def init(path: Optional[Path]) -> None:
+
+    repo = Repository()
+    if repo.active_stack:
+        metadata_store_path = (
```
As any stack component now has an optional `local_path` property, is this something we could maybe extend to not only handle the artifact/metadata store here?

strickvl replied:
I don't understand this comment. Are you suggesting that I could leverage the `local_path` property as a way of discovering whether a component was local or not?

What does 'not only handle' mean?

schustmi replied:
Sorry, could have been clearer on that. Any `StackComponent` can specify a `local_path` that is uses to store persistent information on disk. If we call `zenml clean --local` it would currently delete the content of the metadata and artifact store right? We could extend the functionality to also clean the local directories of other stack components, for example a feature store or container registry.

schustmi replied:
Maybe also good to loop in @stefannica here to see if he thinks this makes sense or what was intended with the `--local` option

stefannica replied:
I agree with @schustmi, you can loop through all the active stack components and, if you find `local_path` set to a non-empty value, delete that path (file or directory(.

strickvl commented on src/zenml/cli/base.py:
```diff
@@ -63,19 +66,116 @@ def init(path: Optional[Path]) -> None:
+    if not force_delete:
+        confirm = confirmation(
+            "DANGER: This will completely delete anything inside the folders for the following stack components: \n"
+            "- local metadata store \n"
```
I don't understand this comment.

htahir1 replied:
`local metadata store` is called `default` now I think?

schustmi reviewed (approved):
LGTM!
